### PR TITLE
Method to get HBase client stats

### PR DESCRIPTION
**Pull request by vshlapakov**
We want to provide timings per each HBase call to message bus, to be later converted to time-series metrics. This PR implements HBase client logic to store last calls and its timestamps.

### CODE REVIEW

File: happybase/client.py [hunk 1]
```diff
@@ -28,41 +43,70 @@ def _req(self, _api, *args, **kwargs):
...
     def _log_response_times(self):
         logger.debug("client id %d response times (ms) statistics", id(self))
-        logger.debug("mean: %f stdev: %f median: %f count: %d",
-                     np.mean(self._resp_times),
-                     np.std(self._resp_times),
-                     np.median(self._resp_times),
-                     len(self._resp_times))
\ No newline at end of file
+        for api_entry in self._response_timings:
+            logger.debug("%s mean: %f stdev: %f median: %f count: %d",
```
Comment (vshlapakov) [inline happybase/client.py]: Having some doubts about this, it's more relevant to print stats per function type, not summary of all values, what do you think? At the same point it means more logs, wonder if it's appropriate.

File: happybase/client.py [hunk 2]
```diff
@@ -1,25 +1,40 @@
...
 
 logger = logging.getLogger(__name__)
 
 
+TIMINGS_KEEP_LIMIT = 50
+EPOCH_DATETIME = datetime.datetime.utcfromtimestamp(0)
+
+
+def utc_timestamp():
+    return (datetime.datetime.utcnow() - EPOCH_DATETIME).total_seconds()
```
Comment (vshlapakov) [inline happybase/client.py]: Before this PR `time.time()` was used everywhere in the module, but as we need to send some timestamps for the metrics to create time-series, it's better/simpler to use UTC timestamps from the beginning.

File: happybase/client.py [hunk 3]
```diff
@@ -28,41 +43,70 @@ def _req(self, _api, *args, **kwargs):
...
+
+    def get_stats(self):
+        """Get HBase client stats.
+
+        Current implementation returns only response timings per each client
+        function, only new values stored after last get_stats() call.
+        """
+        stats, now = {}, utc_timestamp()
+        for api_entry in self._response_timings:
+            stats_per_entry = [(ts, value) for ts, value
```
Comment (vshlapakov) [inline happybase/client.py]: Main point of the PR: we provide timings associated with timestamps for each api function, say:
```
{"callA": [(1499761168, 1), (1499761170, 2), ...], "callB": [...], ...}
```

File: happybase/client.py (hunk 2 above)
Comment (sibiryakov) [reply happybase/client.py]: > time.mktime(datetime.datetime.utcnow().utctimetuple())

File: happybase/client.py [hunk 4]
```diff
@@ -28,41 +43,70 @@ def _req(self, _api, *args, **kwargs):
...
+                    client = self._reopen_connection()
+                    if client:
                         break
                 if no_retry:
                     raise exc
             finally:
-                now = time()
-                elapsed = (now - t) * 1E+3
+                now = utc_timestamp()
+                elapsed = int((now - t) * 1E+3)
```
Comment (sibiryakov) [inline happybase/client.py]: is this really needed? no way to send floats?

File: happybase/client.py [hunk 5]
```diff
@@ -28,41 +43,70 @@ def _req(self, _api, *args, **kwargs):
...
+        """Reopen connection to HBase.
+
+        Method should return a client instance or None on failed attempt.
+        """
+        logger.info("Trying to reconnect")
+        try:
+            self._connection._refresh_thrift_client()
+            self._connection.open()
+            client = super(RecoveringClient, self._connection.client)
+            logger.debug("New client is initialized")
```
Comment (sibiryakov) [inline happybase/client.py]: this line could be moved after else clause

File: happybase/client.py (hunk 1 above)
Comment (sibiryakov) [reply happybase/client.py]: let's remove this logging, since we will get better stats in graphana soon

File: happybase/client.py [hunk 6]
```diff
@@ -1,25 +1,40 @@
...
-from collections import deque
-from time import sleep, time
-import logging
-import numpy as np
+
 
 logger = logging.getLogger(__name__)
 
 
+TIMINGS_KEEP_LIMIT = 50
```
Comment (sibiryakov) [inline happybase/client.py]: this number should be increased probably, at least states flush will not fit

Comment (vshlapakov) [reply happybase/client.py]: Increased up to 200 per AP entry, should be enough, as we're going to send stats per minute for now.

File: happybase/client.py (hunk 4 above)
Comment (vshlapakov) [reply happybase/client.py]: Not really, you're right, we do need floats for timestamps.
Don't think we need floats for elapsed though, it's already in milliseconds.

File: happybase/client.py (hunk 2 above)
Comment (vshlapakov) [reply happybase/client.py]: We need milliseconds to get elapsed time, and `time.struct_time` ignores it

**Pull request merged**